api/tests/authentication_tests.py

```diff
@@ -18,6 +32,213 @@ def _setup_fxa_response(status_code: int, json: dict | str):
+        body=Timeout(),
+        status=status_code,
+        json=json,
+    )
+    responses.add(resp)
+    return {"status_code": status_code}
+
+
+def _setup_fxa_response_no_json(status_code: int):
+    fxa_verify_path = "https://oauth.stage.mozaws.net/v1/introspect"
+    responses.add(responses.POST, fxa_verify_path, status=status_code)
+    return {"status_code": status_code}
+
+
+class AuthenticationMiscellaneous(TestCase):
+    def setUp(self):
+        self.auth = FxaTokenAuthentication
+        self.factory = RequestFactory()
+        self.path = "/api/v1/relayaddresses"
+        self.fxa_verify_path = "https://oauth.stage.mozaws.net/v1/introspect"
+        self.uid = "relay-user-fxa-uid"
+
+    def tearDown(self):
+        cache.clear()
+
+    @responses.activate()
+    def test_introspect_token_catches_JSONDecodeError_raises_AuthenticationFailed(self):
+        _setup_fxa_response_no_json(200)
+        invalid_token = "invalid-123"
+        cache_key = get_cache_key(invalid_token)
+
+        try:
+            introspect_token(cache_key, invalid_token)
+        except AuthenticationFailed as e:
+            assert str(e.detail) == "JSONDecodeError from FXA introspect response"
+            assert responses.assert_call_count(self.fxa_verify_path, 1) is True
+            return
+        self.fail("Should have raised AuthenticationFailed")
+
+    @responses.activate()
+    def test_introspect_token_returns_fxa_introspect_response(self):
+        now_time = int(datetime.now().timestamp())
+        # Note: FXA iat and exp are timestamps in *milliseconds*
+        exp_time = (now_time + 60 * 60) * 1000
+        json_data = {"active": True, "sub": self.uid, "exp": exp_time}
+        status_code = 200
+        expected_fxa_resp_data = {"status_code": status_code, "json": json_data}
+        _setup_fxa_response(status_code, json_data)
+        valid_token = "valid-123"
+        cache_key = get_cache_key(valid_token)
+
+        assert cache.get(cache_key) is None
+
+        fxa_resp_data = introspect_token(cache_key, valid_token)
+        assert responses.assert_call_count(self.fxa_verify_path, 1) is True
+        assert fxa_resp_data == expected_fxa_resp_data
+
+    @responses.activate()
+    def test_get_fxa_uid_from_oauth_token_returns_cached_response(self):
+        user_token = "user-123"
+        now_time = int(datetime.now().timestamp())
+        # Note: FXA iat and exp are timestamps in *milliseconds*
+        exp_time = (now_time + 60 * 60) * 1000
+        fxa_response = _setup_fxa_response(
+            200, {"active": True, "sub": self.uid, "exp": exp_time}
+        )
+        cache_key = get_cache_key(user_token)
+
+        assert cache.get(cache_key) is None
+
+        # get FxA uid for the first time
+        fxa_uid = get_fxa_uid_from_oauth_token(user_token)
+        assert fxa_uid == self.uid
+        assert responses.assert_call_count(self.fxa_verify_path, 1) is True
+        assert cache.get(cache_key) == fxa_response
```

Comment:
Nit: should the cache tests just set the cache directly?
```suggestion
        cache_key = get_cache_key(user_token)
        cache.set(get_cache_key(user_token), fxa_response)
```

Reply:
Intentionally kept the tests like this because I want the test to fail if the `introspect_token` logic changes. I am not sure how we feel about this type of "end-to-end" unit tests. Maybe we can revisit this here: https://mozilla-hub.atlassian.net/browse/MPP-3157